### PR TITLE
Enable simplex 2d noise for Synthetic dataset

### LINKED ISSUE #1027: [Task]: Enabling Unsupervised Workflow: Propose a new thresholding approach
### Anomaly Detection
Anomaly detection is the process of identifying data points, events, or observations within a dataset that significantly deviate from the normal or expected behavior. Anomalies may be caused by a variety of factors, including failures, fraud, or unusual behavior.

Anomaly detection is an essential task in numerous industries, including cybersecurity, finance, healthcare, and manufacturing. It can be used to detect fraud in financial transactions, to identify anomalies in medical data for early disease diagnosis, to detect flaws in manufacturing processes, and to monitor traffic for security threats.

### Anomalib
Anomalib is a deep learning library that aims to collect the best anomaly detection algorithms for testing on both public and private datasets. Anomalib offers several ready-to-use implementations of anomaly detection algorithms described in recent research, as well as a set of tools that make it easier to build and use custom models. The library has a strong focus on image-based anomaly detection, where the goal of the algorithm is to identify anomalous images or anomalous pixel regions within images in a dataset.

## The thresholding problem

Anomaly detection models in Anomalib are trained only on normal images. During inference, the models are tasked with distinguishing anomalous samples from normal samples. The task is similar to a classical binary classification problem, but instead of generating a class label and a confidence score, Anomalib models generate an anomaly score, which quantifies the distance of the sample to the distribution of normal samples seen during training. The range of possible anomaly score values is unbounded and may differ widely between models and/or datasets, which makes it challenging to set a good threshold for mapping the raw anomaly scores to a binary class label (normal vs. anomalous).

## Desired solution
Anomalib currently has an [adaptive thresholding mechanism](https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/utils/metrics/anomaly_score_threshold.py) in place which aims to address the thresholding problem. The adaptive thresholding mechanism computes the F1 score over a validation set for a range of thresholds. The final threshold value is the threshold value that results in the highest F1 score. A major drawback of this approach is that the validation set is required to contain anomalous samples, which might not always be available in real world anomaly detection problems. 

**The goal of this hackathon is to design a fully unsupervised thresholding mechanism that does not rely on anomalous samples.** 

#### Possible approaches
- One possibility is to generate fake anomalous images for the validation set, and then use Anomalib's existing adaptive thresholding mechanism. A simple version of this solution, which makes use of random Perlin noise masks, is already [implemented](https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/data/synthetic.py) in Anomalib. Different noise methods or generative models could possibly improve the results of this method.
- Use the characteristics of the anomaly score distribution of the normal training images to set a confidence interval.
- Other, better, solutions may exist. Be creative!

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by WenjingKangIntel**
# Description

- Provide a summary of the modification as well as the issue that has been resolved. List any dependencies that this modification necessitates.
This enhances the current synthetic dataset noise generation method with simplex 2d noise:
references:
https://en.wikipedia.org/wiki/Simplex_noise
https://github.com/lmas/opensimplex

Need to discuss whether the init function of AnomalibDataModule is proper. if proper, needs to enable other dataset other than mvtec. 

Some other thoughts:
We may need add some APIs to determine mixing ratios of two different types of noises, i.e. perlin2d and simplex2d, instead of using only one type of noise.

- Fixes #1027 

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the [pre-commit style and check guidelines](https://github.com/openvinotoolkit/anomalib/blob/main/CONTRIBUTING.md) of this project.
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes
- [ ] I have added a summary of my changes to the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) (not for minor changes, docs and tests).

### CODE REVIEW

**Review by ashwinvaidya17 — COMMENTED** [review]
This will be very useful. I am fine with the design but maybe @djdameln might have some thoughts about this. There is one issue with circular import though when I try to run the code.

File: src/anomalib/data/noise.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+from enum import Enum
```
Comment (ashwinvaidya17) [inline src/anomalib/data/noise.py:1]: Can you add license and docstring at the top? Also, how about naming this to `noise_type.py` similar to `task_type.py`

**Review by ashwinvaidya17 — COMMENTED** [review]
Thanks for updating this. I am almost fine with the changes.

File: src/anomalib/data/utils/generators/opensimplex.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+from enum import Enum
+from math import floor
+
+import numpy as np
+import torch
+from torch import Tensor
+
+try:
+    from numba import njit, prange
```
[the remark this answers is not in the document]
Comment (WenjingKangIntel) [reply src/anomalib/data/utils/generators/opensimplex.py]: `numba` should be added into requirements, otherwise it would be very slow 🤦‍♀️...

File: src/anomalib/data/utils/generators/opensimplex.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
+"""Helper functions for generating Open Simplex noise."""
+
+### add license
```
[the remark this answers is not in the document]
Comment (WenjingKangIntel) [reply src/anomalib/data/utils/generators/opensimplex.py]: using the original library. thus resolved this.

File: src/anomalib/data/noise.py (hunk 1 above)
Comment (WenjingKangIntel) [reply src/anomalib/data/noise.py:1]: fixed.

**Review by djdameln — APPROVED** [review]
Thanks, this is a simple yet useful extension of the random noise augmentation mechanism. We might need to make some changes to the configuration part in the config.yaml to ensure compatibility with other hackathon submissions, but we will sort this out on our side.
